src/web/crate_details.rs

```diff
@@ -46,72 +46,61 @@ pub struct CrateDetails {
     documentation_url: Option<String>,
 }
 
-impl ToJson for CrateDetails {
-    fn to_json(&self) -> Json {
-        let mut m: BTreeMap<String, Json> = BTreeMap::new();
-        m.insert("name".to_string(), self.name.to_json());
-        m.insert("version".to_string(), self.version.to_json());
-        m.insert("description".to_string(), self.description.to_json());
-        m.insert("authors".to_string(), self.authors.to_json());
-        m.insert("owners".to_string(), self.owners.to_json());
-        m.insert("authors_json".to_string(), self.authors_json.to_json());
-        m.insert("dependencies".to_string(), self.dependencies.to_json());
+impl Serialize for CrateDetails {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("CrateDetails", 28)?;
+        state.serialize_field("metadata", &self.metadata)?;
+        state.serialize_field("name", &self.name)?;
+        state.serialize_field("version", &self.version)?;
+        state.serialize_field("description", &self.description)?;
+        state.serialize_field("authors", &self.authors)?;
+        state.serialize_field("owners", &self.owners)?;
+        state.serialize_field("authors_json", &self.authors_json)?;
+        state.serialize_field("dependencies", &self.dependencies)?;
+
         if let Some(ref readme) = self.readme {
-            m.insert("readme".to_string(), render_markdown(&readme).to_json());
+            state.serialize_field("readme", &render_markdown(&readme))?;
         }
+
         if let Some(ref rustdoc) = self.rustdoc {
-            m.insert("rustdoc".to_string(), render_markdown(&rustdoc).to_json());
+            state.serialize_field("rustdoc", &render_markdown(&rustdoc))?;
         }
```

Comment:
It looks like `rustdoc` and `readme` will sometimes not be serialized. In that case, the `len` argument you pass to `serialize_struct` will be incorrect. Do you know what the purpose of that parameter is?

As an alternative, you could do the `if let Some` check ahead of time and increment/decrement the count if they are/are not present.

Reply:
This is a direct translation of what was, I was planning to cover making it sane in another PR

Reply:
This is not a direct translation because the `len` parameter has no equivalent in `rustc_serialize`. We should make sure we're serializing correctly and not breaking any of `serde`'s internal invariants.

Now that I think of it, it would be great to have tests both with and without `readme` and `rustdoc`.

Reply:
Slightly misunderstood you but I've fixed it